lib/json_rpc/channel.ex

```diff
@@ -11,19 +10,28 @@ if Code.ensure_loaded?(Phoenix.Channel) do
     def handle_in("jsonrpc", message, socket) do
       message
       |> Parser.parse()
-      |> handle_message(socket.assigns.connection_params, socket.assigns.registry)
+      |> handle_message(socket.assigns.connection_params, socket.assigns.registry, socket.assigns.timeout)
       |> case do
         [_, _] = responses -> {:reply, {:ok, Enum.reject(responses, &is_nil/1)}, socket}
         %{error: nil} = response -> {:reply, {:ok, response}, socket}
         %{result: nil} = response -> {:reply, {:error, response}, socket}
       end
     end
 
-    defp handle_message(requests, params, registry) when is_list(requests) do
-      Enum.map(requests, &handle_message(&1, params, registry))
+    def handle_message(requests, params, registry, timeout) when is_list(requests) do
+      Task.Supervisor.async_stream_nolink(
+        JSONRPC.TaskSupervisor,
+        requests,
+        __MODULE__,
+        :handle_message,
+        [params, registry],
+        [ordered: true, timeout: timeout, on_timeout: :kill_task]
+      )
+      |> Enum.zip(requests)
```

Comment:
I would love to be confortable enough with hand crafting supervisors and monitors and such that I could offer an alterative, but I'm not yet; and even if I were I'm not sure this isn't the best option anyway.

Reply:
This seems right to me. We want to start a supervised, unlinked task to apply the function to each element, then do something with each result as it completes. That's exactly what`Task.Supervisor.async_stream_nolink` is for.

Reply:
I think the comment was based on my conversation in slack where I was annoyed that I had to keep it `ordered: true` so it could be associated back to the original request so we could have the original id when there is an error state.
After a lot of thinking I've concluded I was annoyed for no good reason and that's just the way it has to be.